File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/component/SaveFavoritesFlow.kt
package com.mbta.tid.mbta_app.android.component

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.RectangleShape
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.fromHtml
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import com.google.accompanist.permissions.ExperimentalPermissionsApi
import com.mbta.tid.mbta_app.android.ModalRoutes
import com.mbta.tid.mbta_app.android.MyApplicationTheme
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.state.getGlobalData
import com.mbta.tid.mbta_app.android.util.SettingsCache
import com.mbta.tid.mbta_app.android.util.Typography
import com.mbta.tid.mbta_app.android.util.fromHex
import com.mbta.tid.mbta_app.android.util.getLabels
import com.mbta.tid.mbta_app.android.util.notificationPermissionState
import com.mbta.tid.mbta_app.model.Direction
import com.mbta.tid.mbta_app.model.FavoriteSettings
import com.mbta.tid.mbta_app.model.LineOrRoute
import com.mbta.tid.mbta_app.model.RouteStopDirection
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.Stop
import com.mbta.tid.mbta_app.repositories.Settings
import com.mbta.tid.mbta_app.usecases.EditFavoritesContext
import com.mbta.tid.mbta_app.utils.TestData
import com.mbta.tid.mbta_app.viewModel.IToastViewModel
import com.mbta.tid.mbta_app.viewModel.ToastViewModel
import org.koin.compose.koinInject

@Composable
fun SaveFavoritesFlow(
    lineOrRoute: LineOrRoute,
    stop: Stop,
    directions: List<Direction>,
    selectedDirection: Int,
    context: EditFavoritesContext,
    toastViewModel: IToastViewModel = koinInject(),
    isFavorite: (routeStopDirection: RouteStopDirection) -> Boolean,
    updateFavorites: (Map<RouteStopDirection, FavoriteSettings?>) -> Unit,
    onClose: () -> Unit,
    openModal: (ModalRoutes) -> Unit,
) {
    val notificationsFlag = SettingsCache.get(Settings.Notifications)
    if (notificationsFlag) {
        openModal(ModalRoutes.SaveFavorite(lineOrRoute.id, stop.id, selectedDirection, context))
        onClose()
        return
    }

    val global = getGlobalData("SaveFavoritesFlow")
    val isUnFavoriting =
        directions.any { it.id == selectedDirection } &&
            isFavorite(RouteStopDirection(lineOrRoute.id, stop.id, selectedDirection))

    val isBusOneDirection = directions.size == 1 && lineOrRoute.sortRoute.type == RouteType.BUS
    val localContext = LocalContext.current

    fun updateAndToast(update: Map<RouteStopDirection, FavoriteSettings?>) {
        updateFavorites(update)
        val favorited = update.filter { it.value != null }
        val firstFavorite = favorited.entries.firstOrNull() ?: return
        val labels = firstFavorite.key.getLabels(global, localContext)
        var toastText: String? = null

        // If there's only a single favorite, show direction, route, and stop in the toast
        if (favorited.size == 1) {
            toastText =
                labels?.let {
                    localContext.getString(
                        R.string.favorites_toast_add,
                        it.direction,
                        it.route,
                        it.stop,
                    )
                } ?: localContext.getString(R.string.favorites_toast_add_fallback)
        }
        // If there are two favorites and they both have the same route and stop, omit direction
        else if (
            favorited.size == 2 &&
                favorited.keys.all {
                    it.route == firstFavorite.key.route && it.stop == firstFavorite.key.stop
                }
        ) {
            toastText =
                labels?.let {
                    localContext.getString(R.string.favorites_toast_add_multi, it.route, it.stop)
                } ?: localContext.getString(R.string.favorites_toast_add_fallback)
        }

        toastText?.let {
            toastViewModel.showToast(
                ToastViewModel.Toast(it, duration = ToastViewModel.Duration.Short)
            )
        }
    }

    // Save automatically without confirmation modal
    if (isUnFavoriting || isBusOneDirection && directions.any { it.id == selectedDirection }) {
        val rsd = RouteStopDirection(lineOrRoute.id, stop.id, selectedDirection)
        updateAndToast(mapOf(rsd to if (isFavorite(rsd)) null else FavoriteSettings()))

        onClose()
    } else {
        Dialog(onDismissRequest = onClose) {
            FavoriteConfirmation(
                lineOrRoute,
                stop,
                directions,
                selectedDirection,
                context,
                proposedFavorites =
                    directions.associateBy({ it.id }) {
                        // if selectedDirection and already a favorite, then removing favorite.
                        // if not selected direction and already a favorite, then keep it.
                        val isSelected = it.id == selectedDirection
                        val isExistingFavorite =
                            isFavorite(RouteStopDirection(lineOrRoute.id, stop.id, it.id))
                        val onlyOppositeDirectionServed =
                            directions.size == 1 && it.id != selectedDirection
                        val suggestingFavorite =
                            (isSelected xor isExistingFavorite) ||
                                // If the only direction is the opposite one, mark it as favorite
                                // whether or not it already is
                                onlyOppositeDirectionServed
                        if (suggestingFavorite) FavoriteSettings() else null
                    },
                updateFavorites = ::updateAndToast,
                onClose = onClose,
            )
        }
    }
}

@OptIn(ExperimentalPermissionsApi::class)
@Composable
fun FavoriteConfirmation(
    lineOrRoute: LineOrRoute,
    stop: Stop,
    directions: List<Direction>,
    selectedDirection: Int,
    context: EditFavoritesContext,
    proposedFavorites: Map<Int, FavoriteSettings?>,
    updateFavorites: (Map<RouteStopDirection, FavoriteSettings?>) -> Unit,
    onClose: () -> Unit,
) {
    val notificationsFlag = SettingsCache.get(Settings.Notifications)

    val notificationPermissionState = notificationPermissionState()

    var favoritesToSave: Map<Int, FavoriteSettings?> by remember {
        mutableStateOf(proposedFavorites)
    }

    fun saveAndClose() {
        if (notificationsFlag) notificationPermissionState.launchPermissionRequest()
        val newFavorites =
            favoritesToSave.mapKeys { (directionId, _isFavorite) ->
                RouteStopDirection(lineOrRoute.id, stop.id, directionId)
            }
        updateFavorites(newFavorites)
        onClose()
    }

    Column(
        modifier =
            Modifier.background(colorResource(R.color.fill1), shape = RoundedCornerShape(28.dp))
    ) {
        val headerText =
            if (context == EditFavoritesContext.Favorites) {
                stringResource(R.string.add_to_favorites_title, lineOrRoute.name, stop.name)
            } else {
                stringResource(
                    R.string.add_to_favorites_title_stop_details,
                    lineOrRoute.name,
                    stop.name,
                )
            }

        Text(
            AnnotatedString.fromHtml(headerText),
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(16.dp).fillMaxWidth().semantics { heading() },
        )
        if (directions.size == 1 && directions.first().id != selectedDirection) {

            Text(
                stringResource(
                    R.string.direction_service_only,
                    stringResource(directionNameFormatted(directions.first())),
                ),
                textAlign = TextAlign.Center,
                style = Typography.footnoteSemibold,
                modifier =
                    Modifier.padding(horizontal = 16.dp).padding(bottom = 8.dp).fillMaxWidth(),
            )
        }

        if (directions.isEmpty()) {
            Column {
                Text(
                    stringResource(R.string.this_stop_is_drop_off_only),
                    textAlign = TextAlign.Center,
                    style = Typography.footnoteSemibold,
                    modifier =
                        Modifier.padding(horizontal = 16.dp).padding(bottom = 8.dp).fillMaxWidth(),
                )
                HaloSeparator()
                Row(modifier = Modifier.padding(16.dp)) {
                    Spacer(Modifier.weight(1F))
                    TextButton(onClose) { Text(stringResource(R.string.okay)) }
                }
            }
        } else {
            Column() {
                HaloSeparator()
                directions.mapIndexed { idx, direction ->
                    Button(
                        onClick = {
                            val oldValue = favoritesToSave.getOrDefault(direction.id, null)
                            val newValue = if (oldValue == null) FavoriteSettings() else null
                            favoritesToSave = favoritesToSave.plus(direction.id to newValue)
                        },
                        colors =
                            ButtonDefaults.buttonColors(
                                containerColor = colorResource(R.color.fill3),
                                contentColor = colorResource(R.color.text),
                            ),
                        shape = RectangleShape,
                        modifier =
                            Modifier.background(color = colorResource(R.color.fill3)).semantics {
                                selected = favoritesToSave.getOrDefault(direction.id, null) != null
                            },
                    ) {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween,
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            DirectionLabel(direction, Modifier.weight(1f))
                            StarIcon(
                                favoritesToSave.getOrDefault(direction.id, null) != null,
                                color = Color.fromHex(lineOrRoute.backgroundColor),
                                Modifier.padding(start = 16.dp),
                            )
                        }
                    }
                    HaloSeparator()
                }
            }

            Row(modifier = Modifier.padding(16.dp)) {
                Spacer(Modifier.weight(1F))
                TextButton(onClose) { Text(stringResource(R.string.cancel)) }
                TextButton(
                    { saveAndClose() },
                    enabled = notificationsFlag || favoritesToSave.values.any { it != null },
                ) {
                    Text(stringResource(R.string.add_confirmation_button))
                }
            }
        }
    }
}

class Previews() {
    val objects = TestData

    @Preview(name = "Favorites both directions available")
    @Composable
    fun FavoriteConfirmationBoth() {
        MyApplicationTheme {
            Column(Modifier.background(colorResource(R.color.fill3))) {
                FavoriteConfirmation(
                    LineOrRoute.Line(objects.getLine("line-Green"), emptySet()),
                    stop = objects.getStop("place-boyls"),
                    directions =
                        listOf(
                            Direction(id = 0, name = "West", destination = "Copley & West"),
                            Direction(id = 1, name = "East", destination = "Park St & North"),
                        ),
                    selectedDirection = 0,
                    context = EditFavoritesContext.StopDetails,
                    proposedFavorites = mapOf(0 to FavoriteSettings(), 1 to null),
                    updateFavorites = {},
                    onClose = {},
                )
            }
        }
    }

    @Preview(name = "Favorites selected direction available")
    @Composable
    fun FavoriteConfirmationOnlySelectedDirection() {
        MyApplicationTheme {
            Column(Modifier.background(colorResource(R.color.fill3))) {
                FavoriteConfirmation(
                    LineOrRoute.Line(objects.getLine("line-Green"), emptySet()),
                    stop = objects.getStop("place-unsqu"),
                    directions =
                        listOf(Direction(id = 0, name = "West", destination = "Copley & West")),
                    selectedDirection = 0,
                    context = EditFavoritesContext.StopDetails,
                    proposedFavorites = mapOf(0 to FavoriteSettings()),
                    updateFavorites = {},
                    onClose = {},
                )
            }
        }
    }

    @Preview(name = "Favorites opposite direction available")
    @Composable
    fun FavoriteConfirmationOnlyOppositeDirection() {
        MyApplicationTheme {
            Column(Modifier.background(colorResource(R.color.fill3))) {
                FavoriteConfirmation(
                    LineOrRoute.Line(objects.getLine("line-Green"), emptySet()),
                    stop = objects.getStop("place-unsqu"),
                    directions =
                        listOf(Direction(id = 0, name = "West", destination = "Copley & West")),
                    selectedDirection = 1,
                    context = EditFavoritesContext.Favorites,
                    proposedFavorites = mapOf(0 to FavoriteSettings()),
                    updateFavorites = {},
                    onClose = {},
                )
            }
        }
    }

    @Preview(name = "Favorites none added button disables")
    @Composable
    fun FavoriteConfirmationDisabledAdd() {
        MyApplicationTheme {
            Column(Modifier.background(colorResource(R.color.fill3))) {
                FavoriteConfirmation(
                    LineOrRoute.Line(objects.getLine("line-Green"), emptySet()),
                    stop = objects.getStop("place-boyls"),
                    directions =
                        listOf(
                            Direction(id = 0, name = "West", destination = "Copley & West"),
                            Direction(id = 1, name = "East", destination = "Park St & North"),
                        ),
                    selectedDirection = 1,
                    context = EditFavoritesContext.Favorites,
                    proposedFavorites = mapOf(0 to null, 1 to null),
                    updateFavorites = {},
                    onClose = {},
                )
            }
        }
    }
}

File: iosApp/iosApp/ComponentViews/SaveFavoritesFlow.swift
//
//  SaveFavoritesFlow.swift
//  iosApp
//
//  Created by Kayla Brady on 6/24/25.
//  Copyright © 2025 MBTA. All rights reserved.
//
import Collections
import CustomAlert
import Foundation
import Shared
import SwiftUI

enum SaveFavoritesContext {
    case favorites
    case stopDetails
}

struct SaveFavoritesFlow: View {
    let lineOrRoute: LineOrRoute
    let stop: Stop
    let directions: [Direction]
    let selectedDirection: Int32
    let context: SaveFavoritesContext
    let global: GlobalResponse?
    let isFavorite: (RouteStopDirection) -> Bool
    let updateFavorites: ([RouteStopDirection: Bool]) -> Void
    let onClose: () -> Void
    let toastVM: IToastViewModel

    var selectedDirectionIsAvailableAtStop: Bool { directions.contains(where: { $0.id == selectedDirection }) }

    var isUnFavoriting: Bool {
        selectedDirectionIsAvailableAtStop &&
            isFavorite(RouteStopDirection(route: lineOrRoute.id, stop: stop.id, direction: selectedDirection))
    }

    var isBusOneDirection: Bool { directions.count == 1 && lineOrRoute.type == RouteType.bus }

    let inspection = Inspection<Self>()

    init(
        lineOrRoute: LineOrRoute, stop: Stop, directions: [Direction], selectedDirection: Int32,
        context: SaveFavoritesContext, global: GlobalResponse?,
        isFavorite: @escaping (RouteStopDirection) -> Bool,
        updateFavorites: @escaping ([RouteStopDirection: Bool]) -> Void,
        onClose: @escaping () -> Void,
        toastVM: IToastViewModel = ViewModelDI().toast
    ) {
        self.lineOrRoute = lineOrRoute
        self.stop = stop
        self.directions = directions
        self.selectedDirection = selectedDirection
        self.context = context
        self.global = global
        self.isFavorite = isFavorite
        self.updateFavorites = updateFavorites
        self.onClose = onClose
        self.toastVM = toastVM
    }

    func proposedFavorites() -> [Int32: Bool] {
        directions
            .reduce(into: [Int32: Bool]()) { acc, direction in
                // if selectedDirection and already a favorite, then removing favorite.
                // if this is not the selected direction and already a favorite, then keep it.
                let isFavorite = ((direction.id == selectedDirection) !=
                    isFavorite(RouteStopDirection(
                        route: lineOrRoute.id,
                        stop: stop.id,
                        direction: direction.id
                    ))) ||
                    // If the only direction is the opposite one, mark it as favorite whether or not it already is
                    (directions.count == 1 && direction.id != selectedDirection)
                acc[direction.id] = isFavorite
            }
    }

    func updateAndToast(_ update: [RouteStopDirection: Bool]) {
        updateFavorites(update)

        let favorited = update.filter(\.value)
        let firstFavorite = favorited.first
        let labels = firstFavorite?.key.getLabels(global)
        var toastText: String? = nil

        // If there's only a single favorite, show direction, route, and stop in the toast
        if favorited.count == 1 {
            toastText = if let labels {
                String(format: NSLocalizedString(
                    "**%1$@ %2$@** at **%3$@** added to Favorites",
                    comment: """
                    Favorite added toast text, the first value is the direction (southbound, inbound, etc),
                    the second is the route name (Red Line, 1 bus), and the third is a stop name (Ruggles, Alewife).
                    The asterisks surround bolded text. ex. "[Southbound] [Red Line] at [Alewife] added to Favorites"
                    """
                ), labels.direction, labels.route, labels.stop)
            } else {
                NSLocalizedString(
                    "Added to Favorites",
                    comment: "Favorites added toast fallback text when more details are unavailable"
                )
            }
        }
        // If there are two favorites and they both have the same route and stop, omit direction
        else if favorited.count == 2,
                favorited.keys.allSatisfy({
                    $0.route == firstFavorite?.key.route
                        && $0.stop == firstFavorite?.key.stop
                }) {
            toastText = if let labels {
                String(format: NSLocalizedString(
                    "**%1$@** at **%2$@** added to Favorites",
                    comment: """
                    Favorite added toast text, the first value is the route name (Red Line, 1 bus),
                    and the second is a stop name (Ruggles, Alewife).
                    The asterisks surround bolded text. ex. "[Green Line] at [Arlington] added to Favorites"
                    """
                ), labels.route, labels.stop)
            } else {
                NSLocalizedString(
                    "Added to Favorites",
                    comment: "Favorites added toast fallback text when more details are unavailable"
                )
            }
        }

        if let toastText {
            toastVM.showToast(toast: .init(
                message: toastText,
                duration: .short, isTip: false, action: nil
            ))
        }
    }

    var body: some View {
        // Save automatically without confirmation modal
        VStack(spacing: 0) {
            if isUnFavoriting || isBusOneDirection, selectedDirectionIsAvailableAtStop {
                VStack {}
                    .onAppear {
                        let rsd = RouteStopDirection(route: lineOrRoute.id, stop: stop.id, direction: selectedDirection)
                        updateAndToast([rsd: !isFavorite(rsd)])

                        onClose()
                    }
            } else {
                FavoriteConfirmationDialog(
                    lineOrRoute: lineOrRoute,
                    stop: stop,
                    directions: directions,
                    selectedDirection: selectedDirection,
                    context: context,
                    proposedFavorites: proposedFavorites(),
                    updateFavorites: updateAndToast, onClose: onClose
                )
            }
        }.onReceive(inspection.notice) { inspection.visit(self, $0) }
    }
}

struct FavoriteConfirmationDialog: View {
    let lineOrRoute: LineOrRoute
    let stop: Stop
    let directions: [Direction]
    let selectedDirection: Int32
    let context: SaveFavoritesContext
    let proposedFavorites: [Int32: Bool]
    let updateFavorites: ([RouteStopDirection: Bool]) -> Void
    let onClose: () -> Void

    @State var showDialog = false
    @State var favoritesToSave: [Direction: Bool] = [:]

    var body: some View {
        VStack {}
            .onAppear {
                favoritesToSave = directions.reduce(into: [Direction: Bool]()) { acc, direction in
                    acc[direction] = proposedFavorites[direction.id] ?? false
                }
                showDialog = true
            }
            .customAlert(
                isPresented: $showDialog,
                content: {
                    FavoriteConfirmationDialogContents(
                        lineOrRoute: lineOrRoute,
                        stop: stop, directions: directions,
                        selectedDirection: selectedDirection,
                        context: context,
                        favoritesToSave: favoritesToSave,
                        updateLocalFavorite: { direction, isFavorite in
                            favoritesToSave[direction] = isFavorite
                        }
                    )
                },
                actions: {
                    if directions.isEmpty {
                        Button {
                            onClose()
                        } label: {
                            Text("Okay")
                        }
                    } else {
                        MultiButton {
                            FavoriteConfirmationDialogActions(
                                lineOrRoute: lineOrRoute,
                                stop: stop,
                                favoritesToSave: favoritesToSave,
                                updateFavorites: updateFavorites,
                                onClose: onClose
                            )
                        }
                    }
                }
            )
    }
}

struct FavoriteConfirmationDialogContents: View {
    let lineOrRoute: LineOrRoute
    let stop: Stop
    let directions: [Direction]
    let selectedDirection: Int32
    let context: SaveFavoritesContext
    let favoritesToSave: [Direction: Bool]
    let updateLocalFavorite: (Direction, Bool) -> Void

    var body: some View {
        let headerText = if context == SaveFavoritesContext.favorites {
            String(format: NSLocalizedString("Add **%1$@** at **%2$@**",
                                             comment: """
                                             Title for a confirmation modal when a user adds a favorite route + stop \
                                             and already has the context that what they are adding is a favorite. \
                                             Ex: Add [Green Line] at [Boylston]
                                             """), lineOrRoute.name, stop.name)
        } else {
            String(format: NSLocalizedString("Add **%1$@** at **%2$@** to Favorites",
                                             comment: """
                                             Title for a confirmation modal when a user adds a favorite route + stop. \
                                             Ex: Add [Green Line] at [Boylston] to Favorites
                                             """), lineOrRoute.name, stop.name)
        }

        VStack {
            Text(AttributedString.tryMarkdown(headerText))
                .font(Typography.body)
                .accessibilityAddTraits(.isHeader)
                .accessibilityHeading(.h1)
            if directions.count == 1, directions.first!.id != selectedDirection {
                Text("\(DirectionLabel.directionNameFormatted(directions.first!)) service only")
                    .font(Typography.footnoteSemibold)
                    .padding(.horizontal, 16)
                    .padding(.top, 8)
            }

            if directions.isEmpty {
                Text("This stop is drop-off only")
                    .font(Typography.footnoteSemibold)
                    .padding(.horizontal, 16)
                    .padding(.top, 8)
            } else {
                VStack(spacing: 0) {
                    DirectionButtons(lineOrRoute: lineOrRoute,
                                     favorites: favoritesToSave,
                                     updateLocalFavorite: updateLocalFavorite)
                }
            }
        }.accessibilityAddTraits(.isModal)
    }
}

struct DirectionButtons: View {
    let lineOrRoute: LineOrRoute
    let favorites: [Direction: Bool]
    let updateLocalFavorite: (Direction, Bool) -> Void

    /**
     Convenience struct to support iterating through the map of favoritesToSave entries via `ForEach`.
     Cannot use an id of direction.id, as we need to rerender each row when the value of isFavorite changes.
     */
    struct DirectionFavorite: Hashable {
        let direction: Direction
        let isFavorite: Bool
    }

    var directionValues: [DirectionFavorite] {
        favorites
            .map { DirectionFavorite(direction: $0.key, isFavorite: $0.value) }
            .sorted(by: { $0.direction.id < $1.direction.id })
    }

    var body: some View {
        VStack(spacing: 0) {
            ForEach(
                directionValues
                    .enumerated()
                    .sorted(by: { $0.offset < $1.offset }),
                id: \.element.direction
            ) { index, directionValue in
                Button(action: {
                    updateLocalFavorite(directionValue.direction, !directionValue.isFavorite)
                }) {
                    VStack(spacing: 0) {
                        HStack(spacing: 0) {
                            DirectionLabel(direction: directionValue.direction)
                                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
                            Spacer()
                            StarIcon(starred: directionValue.isFavorite, color: .init(hex: lineOrRoute.backgroundColor))
                        }
                        .fixedSize(horizontal: false, vertical: true)
                        .padding(.horizontal, 16)
                        .padding(.vertical, 12)
                        if index < directionValues.count - 1 {
                            HaloSeparator()
                        }
                    }
                }
                .accessibilityAddTraits(directionValue.isFavorite ? [.isSelected] : [])
                .fullFocusSize()
                .background(Color.fill3)
            }
        }.withRoundedBorder()
            .padding(.horizontal, 12)
            .padding(.top, 8)
    }
}

struct FavoriteConfirmationDialogActions: View {
    let lineOrRoute: LineOrRoute
    let stop: Stop
    let favoritesToSave: [Direction: Bool]
    let updateFavorites: ([RouteStopDirection: Bool]) -> Void
    let onClose: () -> Void

    var body: some View {
        Button {
            onClose()
        } label: {
            Text("Cancel")
        }

        Button {
            updateFavorites(favoritesToSave
                .reduce(into: [RouteStopDirection: Bool]()) { partialResult, entry in
                    partialResult[RouteStopDirection(
                        route: lineOrRoute.id,
                        stop: stop.id,
                        direction: entry.key.id
                    )] = entry.value
                })
            onClose()
        } label: {
            Text("Add")
        }.disabled(favoritesToSave.values.allSatisfy { $0 == false })
    }
}
